### ISSUE TITLE
ScheduledJob endpoint returns CronJob types.

### ISSUE THREAD

**Issue by krousey** [kubernetes/kubernetes#38656]
**What keywords did you search in Kubernetes issues before filing this one?**: scheduledjob

---

**Is this a BUG REPORT or FEATURE REQUEST?**: BUG REPORT

**Kubernetes version** (use `kubectl version`):
```console
Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.6", GitCommit:"e569a27d02001e343cb68086bc06d47804f62af6", GitTreeState:"clean", BuildDate:"2016-11-12T05:22:15Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5+", GitVersion:"v1.5.0-beta.3.6+b930bfda10fba6", GitCommit:"b930bfda10fba670bf09ed3936ccdc9724e4c860", GitTreeState:"clean", BuildDate:"2016-12-12T21:02:27Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: GCE
- **OS** (e.g. from /etc/os-release):
```console
$ cat /etc/os-release
BUILD_ID=8977.0.0
NAME="Container-VM Image"
GOOGLE_CRASH_ID=Lakitu
VERSION_ID=56
BUG_REPORT_URL=https://crbug.com/new
PRETTY_NAME="Google Container-VM Image"
VERSION=56
GOOGLE_METRICS_PRODUCT_ID=26
HOME_URL="https://cloud.google.com/compute/docs/containers/vm-image/"
ID=gci
```
- **Kernel** (e.g. `uname -a`): Linux e2e-test-krousey-master 4.4.21+ #1 SMP Thu Nov 10 02:50:15 PST 2016 x86_64 Intel(R) Xeon(R) CPU @ 2.30GHz GenuineIntel GNU/Linux
- **Install tools**: ./hack/test/e2e.go -v -up

**What happened**: ScheduledJob endpoints return CronJob resources after the rename.

**What you expected to happen**: ScheduledJob endpoints should return ScheduledJob resources

**How to reproduce it** (as minimally and precisely as possible): Turn up a 1.5 cluster. curl the /apis/batch/v2alpha1/scheduledjobs endpoint. It returns

```json
{
  "kind": "CronJobList",
  "apiVersion": "batch/v2alpha1",
  "metadata": {
    "selfLink": "/apis/batch/v2alpha1/namespaces/kube-system/scheduledjobs",
    "resourceVersion": "10063"
  },
  "items": null
}
```

cc @kubernetes/sig-api-machinery @deads2k @smarterclayton

**Comment by krousey** [comment]
In the current state, we're releasing an alpha API endpoint that is broken. CronJob still works fine. This is probably a good candidate for a 1.5.1 point release.

**Comment by krousey** [comment]
Found while looking in to https://github.com/kubernetes/kubernetes/issues/38232#issuecomment-266574690

**Comment by xingzhou** [comment]
Anyone working on this? if no, I'd like to work on this

**Comment by soltysh** [comment]
That was the simplest approach we could use to maintain backwards compatibility with SJ, but without introducing co-habitating resources. Generally, CJ are still an alpha feature so I'd like to not invest too much time into fixing not-so-supported feature.

**Comment by krousey** [comment]
I understand that this was done to maintain backwards compatibility, but now we have an API endpoint that just doesn't work. It does not conform to other APIs. If I create a type, I expect to get the same type or a `Status` error as a response.

I think at the very least, `ScheduledJob` should be removed completely. It isn't backwards compatible as it is, and it doesn't need to be since this is all alpha. I also think the API machinery should either be enhanced to let something like this work, or guarded to not let something like this happen.

**Comment by smarterclayton** [comment]
Part of the reason we pushed to do this was to get most of the machinery in
place to allow it.  We'd need to make the decision to continue investing to
get to that next tier of capability.  Partial scheduled job is better than
no scheduled job, and we identified and fixed a number of issues that would
block a real rename.  Only clients that read Kind are broken, but that's
not all clients by far.

On Dec 13, 2016, at 1:15 PM, krousey <notifications@github.com> wrote:

I understand that this was done to maintain backwards compatibility, but
now we have an API endpoint that just doesn't work. It does not conform to
other APIs. If I create a type, I expect to get the same type or a Status
error as a response.

I think at the very least, ScheduledJob should be removed completely. It
isn't backwards compatible as it is, and it doesn't need to be since this
is all alpha. I also think the API machinery should either be enhanced to
let something like this work, or guarded to not let something like this
happen.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/38656#issuecomment-266817032>,
or mute the thread
